### PR TITLE
What's Here for Range

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Also corrects a few minor errors and a major omission (#% never had any doc!).

### CODE REVIEW

File: range.c [hunk 1]
```diff
@@ -554,6 +553,33 @@ range_step(int argc, VALUE *argv, VALUE range)
...
 }
 
+/*
+ *  call-seq:
+ *    %(n) {|element| ... } -> self
+ *    %(n)                  -> enumerator
+ *
+ *  Iterates over the elements of +self+.
+ *
+ *  With a block given and no argument,
```
Comment (jeremyevans) [inline range.c:563]: Examples show an argument is used, and function definition shows an argument is required.  I would just remove ` and no argument`.

File: range.c [hunk 2]
```diff
@@ -554,6 +553,33 @@ range_step(int argc, VALUE *argv, VALUE range)
...
+ *  calls the block each element of the range; returns +self+:
```
Comment (jeremyevans) [inline range.c:564]: The example shows the block is not called with each element of the range.  The example shows the block is called with every nth element in the range, as specified by the argument.  So this should be reworded.

File: range.c [hunk 3]
```diff
@@ -2192,6 +2218,76 @@ range_count(int argc, VALUE *argv, VALUE range)
...
+ *
+ * Here, class \Range provides methods that are useful for:
+ *
+ * - {Creating a Range}[#class-Range-label-Methods+for+Creating+a+Range]
+ * - {Querying}[#class-Range-label-Methods+for+Querying]
+ * - {Searching}[#class-Range-label-Methods+for+Searching]
+ * - {Comparing}[#class-Range-label-Methods+for+Comparing]
+ * - {Iterating}[#class-Range-label-Methods+for+Iterating]
+ * - {Converting}[#class-Range-label-Methods+for+Converting]
+ * - {Handling JSON}[#class-Range-label-Methods+for+Handling+JSON]
```
Comment (jeremyevans) [inline range.c:2237]: The Handling JSON section should be removed.  These methods are not part of the core Range class, they are added by the json library, and not available by default.

File: range.c [hunk 4]
```diff
@@ -2192,6 +2218,76 @@ range_count(int argc, VALUE *argv, VALUE range)
...
+ *
+ * - Inherits from {class Object}[Object.html#class-Object-label-What-27s+Here].
+ * - Includes {module Enumerable}[Enumerable.html#module-Enumerable-label-What-27s+Here],
+ *   which provides dozens of additional methods.
+ *
+ * Here, class \Range provides methods that are useful for:
+ *
+ * - {Creating a Range}[#class-Range-label-Methods+for+Creating+a+Range]
+ * - {Querying}[#class-Range-label-Methods+for+Querying]
+ * - {Searching}[#class-Range-label-Methods+for+Searching]
```
Comment (jeremyevans) [inline range.c:2233]: Searching only has a single entry, `bsearch`, and I think that can be moved to the Querying section.

File: range.c [hunk 5]
```diff
@@ -2192,6 +2218,76 @@ range_count(int argc, VALUE *argv, VALUE range)
...
+ * - {Handling JSON}[#class-Range-label-Methods+for+Handling+JSON]
+ *
+ * === Methods for Creating a \Range
+ *
+ * - ::new:: Returns a new range.
+ *
+ * === Methods for Querying
+ *
+ * - #begin:: Returns the begin value given for +self+.
+ * - #count:: Returns a count of elements in +self+: absolute,
```
Comment (jeremyevans) [inline range.c:2246]: Not sure what `absolute` is supposed to mean here.

File: range.c [hunk 6]
```diff
@@ -2192,6 +2218,76 @@ range_count(int argc, VALUE *argv, VALUE range)
...
+ * === Methods for Comparing
+ *
+ * - {#==}[#method-i-3D-3D]:: Returns whether a given object is equal to +self+
+ *                            (uses #==).
+ * - #===:: Returns whether the given object is between the begin and end values.
+ * - #eql?:: Returns whether a given object is equal to +self+ (uses #eql?).
+ *
+ * === Methods for Iterating
+ *
+ * - #%:: Iterates over selected elements of +self+,
```
Comment (jeremyevans) [inline range.c:2274]: All three entries have the same description, but slightly different behavior.  `step` should indicate it only yields nth elements (defaulting to 1).  `%` should mention it is the same as `step`, but requires an argument.

File: range.c [hunk 7]
```diff
@@ -2192,6 +2218,76 @@ range_count(int argc, VALUE *argv, VALUE range)
...
+ *        passing each to the given block.
+ * - #each:: Iterates over the elements of +self+, passing each to the given block.
+ * - #step:: Iterates over selected elements of +self+,
+ *           passing each to the given block.
+ *
+ * === Methods for Converting
+ *
+ * - #inspect:: Returns a string representation of +self+.
+ * - #to_a (aliased as #entries):: Returns elements of +self+ in an array.
+ * - #to_s:: Returns a string representation of +self+.
```
Comment (jeremyevans) [inline range.c:2284]: Might want to treat inspect/to_s similar to ==/eql?, and mention that inspect uses inspect and to_s uses to_s.

File: range.c [hunk 8]
```diff
@@ -2192,6 +2218,76 @@ range_count(int argc, VALUE *argv, VALUE range)
...
+ *
+ * === Methods for Searching
+ *
+ * - #bsearch:: Returns an element from +self+ selected by a binary search.
+ *
+ * === Methods for Comparing
+ *
+ * - {#==}[#method-i-3D-3D]:: Returns whether a given object is equal to +self+
+ *                            (uses #==).
+ * - #===:: Returns whether the given object is between the begin and end values.
```
Comment (jeremyevans) [inline range.c:2269]: The behavior of `===` is almost the same as `cover?`, so they should be in the same section.

File: range.c [hunk 9]
```diff
@@ -554,6 +553,33 @@ range_step(int argc, VALUE *argv, VALUE range)
...
+ *    a # => [1, 3, 5]
+ *    a = []
+ *    ('a'..'e').%(2) {|element| a.push(element) } # => "a".."e"
+ *    a # => ["a", "c", "e"]
+ *
+ *  With no block given, returns an enumerator,
+ *  which will be of class Enumerator::ArithmeticSequence if +self+ is numeric;
+ *  otherwise of class Enumerator:
+ *
+ *    e = (1..5).%(2) # => ((1..5).%(2))
```
Comment (jeremyevans) [inline range.c:577]: Can we use operator syntax here?: `(1..5) % 2`.  Similar for a couple lines below.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks mostly good, but I do have some requests for changes.

File: range.c (hunk 5 above)
Comment (BurdetteLamar) [reply range.c:2246]: > 
> 
> Not sure what `absolute` is supposed to mean here.
Would "absolute count" or "actual count" be better?

Comment (jeremyevans) [reply range.c:2246]: Maybe just leave it at `Returns a count of elements in +self+`?  Maybe I am missing some of the subtlety added by the additional explanation?

File: range.c (hunk 8 above)
Comment (BurdetteLamar) [reply range.c:2269]: > 
> 
> Which section, please?

NM:  Putting into Comparing.

File: range.c (hunk 1 above)
Comment (BurdetteLamar) [reply range.c]: Done.

File: range.c (hunk 2 above)
Comment (BurdetteLamar) [reply range.c]: Done.

File: range.c (hunk 3 above)
Comment (BurdetteLamar) [reply range.c]: Done.

File: range.c (hunk 4 above)
Comment (BurdetteLamar) [reply range.c]: Done.

File: range.c (hunk 5 above)
Comment (BurdetteLamar) [reply range.c]: Done.

File: range.c (hunk 6 above)
Comment (BurdetteLamar) [reply range.c]: Done..

File: range.c (hunk 7 above)
Comment (BurdetteLamar) [reply range.c]: Done.

File: range.c (hunk 9 above)
Comment (BurdetteLamar) [reply range.c]: Done.

File: range.c [hunk 10]
```diff
@@ -554,6 +553,33 @@ range_step(int argc, VALUE *argv, VALUE range)
...
 }
 
+/*
+ *  call-seq:
+ *    %(n) {|element| ... } -> self
+ *    %(n)                  -> enumerator
+ *
+ *  Iterates over the elements of +self+.
+ *
+ *  With a block given, calls the block selected elements of the range;
```
Comment (jeremyevans) [inline range.c:563]: `block selected` -> `block with selected` ?

File: range.c [hunk 11]
```diff
@@ -2192,6 +2218,64 @@ range_count(int argc, VALUE *argv, VALUE range)
...
+ * === Methods for Querying
+ *
+ * - #begin:: Returns the begin value given for +self+.
+ * - #bsearch:: Returns an element from +self+ selected by a binary search.
+ * - #count:: Returns a count of elements in +self+.
+ * - #end:: Returns the end value given for +self+.
+ * - #exclude_end?:: Returns whether the end object is excluded.
+ * - #first:: Returns the first elements of +self+.
+ * - #hash:: Returns the integer hash code.
+ * - #include? (aliased as #member?):: Returns whether a given object
```
Comment (jeremyevans) [inline range.c:2250]: Since `cover?` and `===` are in Comparing, `include?` should be Comparing as well.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Almost there.  Just a couple more change requests.  After those changes, this can committed without further review.

File: range.c (hunk 11 above)
Comment (BurdetteLamar) [reply range.c]: Done.

File: range.c (hunk 10 above)
Comment (BurdetteLamar) [reply range.c]: Done.

**Comment by JuPlutonic** [comment]
Late, and not for actuall changes, but:
https://github.com/ruby/ruby/blob/3ac85a4b7ead0248b4a30fd605ade32750bdf9da/range.c#L229
I'm thinking what it makes sense only if they not equal <tt>!=</tt>.

**Pull request merged**